Signal/src/util/AppUpdateNag.m

```diff
@@ -0,0 +1,130 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import "AppUpdateNag.h"
+#import "RegistrationViewController.h"
+#import "Signal-Swift.h"
+#import <ATAppUpdater/ATAppUpdater.h>
+#import <SignalServiceKit/TSStorageManager.h>
+
+NSString *const TSStorageManagerAppUpgradeNagCollection = @"TSStorageManagerAppUpgradeNagCollection";
+NSString *const TSStorageManagerAppUpgradeNagDate = @"TSStorageManagerAppUpgradeNagDate";
+
+@interface AppUpdateNag () <ATAppUpdaterDelegate>
+
+@property (nonatomic, readonly) TSStorageManager *storageManager;
+
+@end
+
+#pragma mark -
+
+@implementation AppUpdateNag
+
++ (instancetype)sharedInstance
+{
+    static AppUpdateNag *sharedInstance = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedInstance = [[self alloc] initDefault];
+    });
+    return sharedInstance;
+}
+
+- (instancetype)initDefault
+{
+    TSStorageManager *storageManager = [TSStorageManager sharedManager];
+
+    return [self initWithStorageManager:storageManager];
+}
+
+- (instancetype)initWithStorageManager:(TSStorageManager *)storageManager
+{
+    self = [super init];
+
+    if (!self) {
+        return self;
+    }
+
+    OWSAssert(storageManager);
+
+    _storageManager = storageManager;
+
+    OWSSingletonAssert();
+
+    return self;
+}
+
+- (void)showAppUpgradeNagIfNecessary
+{
+    // Only show nag if we are "at rest" in the home view or registration view without any
+    // alerts or dialogs showing.
+    UIViewController *frontmostViewController =
+    [UIApplication sharedApplication].frontmostViewController;
+    OWSAssert(frontmostViewController);
+    BOOL canPresent = ([frontmostViewController isKindOfClass:[SignalsViewController class]] ||
+                       [frontmostViewController isKindOfClass:[RegistrationViewController class]]);
+    if (!canPresent) {
+        return;
+    }
+    
+    NSDate *lastNagDate = [[TSStorageManager sharedManager] dateForKey:TSStorageManagerAppUpgradeNagDate
+                                                          inCollection:TSStorageManagerAppUpgradeNagCollection];
+    const NSTimeInterval kMinute = 60.f;
+    const NSTimeInterval kHour = 60 * kMinute;
+    const NSTimeInterval kDay = 24 * kHour;
+    const NSTimeInterval kNagFrequency = kDay * 14;
+    BOOL canNag = (!lastNagDate || fabs(lastNagDate.timeIntervalSinceNow) > kNagFrequency);
+    if (!canNag) {
+        return;
+    }
+
+    // NOTE: The iTunes app store API exposes "short" version numbers, so
```

Comment:
Dur, this actually makes sense.  Yes, the iTunes API exposes 3 (not 4) components of the version number.  I was thinking it would be an issue that the fourth (build number) component was missing, but I forgot that iTunes release tracks will only release a single build.  Please ignore!